pipeline/backend/ssh/ssh.go

```diff
@@ -0,0 +1,141 @@
+package ssh
+
+import (
+	"context"
+	"encoding/base64"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/melbahja/goph"
+
+	"github.com/woodpecker-ci/woodpecker/pipeline/backend/types"
+	"github.com/woodpecker-ci/woodpecker/server"
+)
+
+type ssh struct {
+	cmd    *goph.Cmd
+	output io.ReadCloser
+	client *goph.Client
+}
+
+type readCloser struct {
+	io.Reader
+}
+
+func (c readCloser) Close() error {
+	return nil
+}
+
+// make sure local implements Engine
+var _ types.Engine = &ssh{}
+
+// New returns a new ssh Engine.
+func New() types.Engine {
+	return &ssh{}
+}
+
+func (e *ssh) Name() string {
+	return "ssh"
+}
+
+func (e *ssh) IsAvailable() bool {
+	return os.Getenv("WOODPECKER_SSH_KEY") != "" || os.Getenv("WOODPECKER_SSH_PASSWORD") != ""
+}
+
+func (e *ssh) Load() error {
+	var auth goph.Auth
+	if file, has := os.LookupEnv("WOODPECKER_SSH_KEY"); has {
+		var err error
+		auth, err = goph.Key(file, os.Getenv("WOODPECKER_SSH_KEY_PASSWORD"))
+		if err != nil {
+			return err
+		}
+	} else {
+		auth = goph.Password(os.Getenv("WOODPECKER_SSH_PASSWORD"))
+	}
+	client, err := goph.New(os.Getenv("WOODPECKER_SSH_USER"), os.Getenv("WOODPECKER_SSH_HOST"), auth)
+	if err != nil {
+		return err
+	}
+	e.client = client
+	return nil
+}
+
+// Setup the pipeline environment.
+func (e *ssh) Setup(ctx context.Context, proc *types.Config) error {
+	return nil
+}
+
+// Exec the pipeline step.
+func (e *ssh) Exec(ctx context.Context, proc *types.Step) error {
+	// Get environment variables
+	Command := []string{}
+	for a, b := range proc.Environment {
+		if a != "HOME" && a != "SHELL" { // Don't override $HOME and $SHELL
+			Command = append(Command, a+"="+b)
```

Comment:
There should be quotes around b, and b should be properly escaped. Otherwise, a environment variable with space would result in weird failure, and potential code execution (albeit, if you control the environment, there is much easier way):
if PATH has /bin:/usr/local/directory with space/bin, it will be PATH=/bin:/usr/local/directory with space/bin command, eg, executing "with" as a command, and not command. 

I give PATH as a example, but on my current system, I have 3 env variable with spaces:

PERL_MB_OPT=--install_base "/home/user/perl5"
LESSOPEN=||/usr/bin/lesspipe.sh %s
which_declare=declare -f


LESSOPEN is set in /etc/profile.d/less.sh, from less-590-2.fc35.x86_64. 

I never really understood the semantics of profile vs bashrc and I run the agent with systemd, so the environment would be clean, but some people might not, and so might face weird issue due to that.

Reply:
In fact, I wonder why this code is here. For example, this override HOSTNAME with the agent one, which would be surprising. It also override PATH or LANG, both variable that would have side effect on execution (for example, the side effect of translating error messages and/or changing sort order, as my team found out one day).

I am surely missing something, but could the reason for passing variable be explained a bit ?

Reply:
It also seems to leak WOODPECKER_SSH_PASSWORD and WOODPECKER_SSH_KEY_PASSWORD to the remote system. It doesn't seems like a great idea, especially if you are running untrusted pipeline or something.

Reply:
Reading the code a bit more, it seems I was wrong proc.Environment is not the same as os.Environment, so there is no variable leak. I however still think to verify the space issue is valid, since it get passed to a remote shell by ssh (and I think local backend should be ok, but I need to verify)

Reply:
If somebody would like to fix this (would recommend a separate PR since this is not limited to this backend), you can take a look at https://github.com/alessio/shellescape

Reply:
If somebody would like to fix this (would recommend a separate PR since this is not limited to this backend), you can take a look at https://github.com/alessio/shellescape